eip-0012.md

```diff
@@ -0,0 +1,277 @@
+```
+{
+	boxId: BoxId,
+	spendingProof: ProverResult,
+}
+```
+This is `Input` in `sigma-rust`.
+
+## Paginate
+
+```
+{
+	page: int,
+	limit: int,
+}
+```
+Used to specify optional pagination for some API calls. Limits results to {limit} each page, and uses a 0-indexing {page} to refer to which of those pages of {limit} items each.
+
+## PaginateError
+
+```
+{
+	maxSize: int,
+}
+```
+{maxSize} is the maximum size for pagination and if we try to request pages outside of this boundary this error is thrown.
+
+### ProverResult
+Uses `sigma-rust` rep - proof is a byte array serialized using `SigmaSerialize` in hex format:
+```
+{
+	proofBytes: String,
+	extension: ContextExtension,
+}
+```
+
+### SignedTx
+
+Uses `sigma-rust` rep for a transaction:
+
+```
+{
+	id: TxId,
+	inputs: [SignedInput],
+	dataInputs: [Input],
+	outputs: [Box],
+	size: int,
+}
+```
+with the difference with `Tx` is that the inputs are signed (`Input` instead of `DataInput`)
+
+
+### TokenAmount
+
+Uses `sigma-rust` rep:
+```
+{
+	tokenId: TokenId,
+	amount: Value,
+}
+```
+
+### TokenId
+
+Uses `sigma-rust` rep - Hex string of the id.
+
+### Tx
+
+An unsigned transaction. Uses a modified version of the transaction representation in `sigma-rust`.
+```
+{
+	id: TxId,
+	inputs: [Input],
+	dataInputs: [Input],
+	outputCandidates: [BoxCandidate],
+}
+```
+This differs from `Tx` in that that `inputs` would be `[Input]`, however we haven't signed this tx yet and also the outputs are merely candidates.
+
+### TxId
+
+Same as `sigma-rust` - string hex encoding
+
+### TxSendError
+
+```
+TxSendErrorCode {
+	Refused: 1,
+	Failure: 2,
+}
+TxSendError {
+	code: TxSendErrorCode,
+	info: String
+}
+```
+
+* Refused - Wallet refuses to send the tx (could be rate limiting)
+* Failure - Wallet could not send the tx
+
+### TxSignError
+
+```
+TxSignErrorCode {
+	ProofGeneration: 1,
+	UserDeclined: 2,
+}
+TxSignError {
+	code: TxSignErrorCode,
+	info: String
+}
+```
+
+* ProofGeneration - User has accepted the transaction sign, but the wallet was unable to sign the transaction (e.g. not having some of the private keys)
+* UserDeclined - User declined to sign the transaction
+
+
+### DataSignError
+
+```
+DataSignErrorCode {
+	ProofGeneration: 1,
+	AddressNotPK: 2,
+	UserDeclined: 3,
+	InvalidFormat: 4,
+}
+DataSignError {
+	code: DataSignErrorCode,
+	info: String
+}
+```
+
+* ProofGeneration - Wallet could not sign the data (e.g. does not have the secret key associated with the address)
+* AddressNotPK - Address was not a P2PK address and thus had no SK associated with it.
+* UserDeclined - User declined to sign the data
+* InvalidFormat - If a wallet enforces data format requirements, this error signifies that the data did not conform to valid formats.
+
+### Value
+
+BigNum-like object. Represents a value which may or may not be ERG. Value is in the smallest possible unit (e.g. nanoErg for ERG). It can be either a `number` or a `string` using standard unsigned integer text representation.
```

Comment:
Value is a positive but encoded as i64 number actually.

Reply:
@kushti Actually it was a "raw" u64 initially, but now I made a newtype `BoxValue(u64)`

Reply:
If it's positive somewhere (even if Long/i64 encoded), then it should be fine as-is here, right? I didn't look into how Scala handles `Long` deserialization from JSON for numbers too big to fit, but in sigma-rust the wrapped u64 should take in a string-encoded uint like "1053546432154" afaik. Please correct me if I'm wrong and this should be changed.

Reply:
Scala representation of Long is signed, so max value is 2^63-1.

Reply:
For JSON deserialization in the full node is this description fine though? Wouldn't the max possible value for nanoergs be much smaller than this? I guess for other token types it could be relevant - in which case should we keep it as-is but specify that it cant' be more than a 64-bit signed number or what?